### PR TITLE
🎉 Destination-snowflake: start using new S3StreamCopier, and expose the purgeStagingData option

### LINKED ISSUE #8820: Snowflake S3 copy destination should use the updated S3StreamCopier, and expose the purgeStagingData option
## Tell us about the problem you're trying to solve
As of https://github.com/airbytehq/airbyte/pull/8607, the SnowflakeCopyS3Destination relies on the old `LegacyS3StreamCopier`. Similarly to Redshift (https://github.com/airbytehq/airbyte/issues/8550), this should be switched to use the new version of `S3StreamCopier`, which calls out to `destination-s3` under the hood.

`destination-snowflake` should then accept a new `purge_staging_data` option.

## Describe the solution you’d like
1. Update `SnowflakeS3StreamCopier` to extend `S3StreamCopier` and `SnowflakeS3StreamCopierFactory` to extend `S3StreamCopierFactory`.
    1. Make sure to add a unit test that calls `copyStagingFileToTemporaryTable` and checks that it executes the correct `COPY INTO` statement, similarly to the [Redshift test](https://github.com/airbytehq/airbyte/blob/9acfc819dc02f77cc9533fb20b572279d9b60a4a/airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftStreamCopierTest.java#L141-L149).
2. Add a `purge_staging_data` option to snowflake's S3 loading config (e.g. https://github.com/airbytehq/airbyte/pull/8855/files#diff-b31312916e0232455938279d1a6d18ea7abd1d49f2b1ce0b56eaed1f31540f3d )
3. Parse that flag and pass it into the S3CopyConfig (e.g. https://github.com/airbytehq/airbyte/pull/8855/files#diff-a66710c91c2b49af79d1976ae0cf45ea2ce9aa47130d0e99a1bdc836670a9c5b)
4. Update the documentation (e.g. https://github.com/airbytehq/airbyte/pull/8855/files#diff-d896670f41eece9b6b5df7c2bf60ba727e8e9e96f257010a2e826eb92e447250)

### PR DESCRIPTION
**Pull request by etsybaev**
## What
Currently, Snowflake uses Deprecated own s3 writer

## How
Updated connector to use central s3 writer in the same way like we already have for Redshift connector

Selected "dev" image with changes to use:
Selection_399

2 streams selected to migrate from test Postgres source to Snoflake:
Tested on UI locally
Selection_397
Selection_396

Result on Snowflake:
Selection_398

## 🚨 User Impact 🚨
PurgeStagingData option is added

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [x] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [x] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java [hunk 1]
```diff
@@ -28,7 +29,7 @@ public AirbyteMessageConsumer getConsumer(final JsonNode config,
         getDatabase(config),
         getSqlOperations(),
         getNameTransformer(),
-        getS3DestinationConfig(config),
+        new S3CopyConfig(S3CopyConfig.shouldPurgeStagingData(config), getS3DestinationConfig(config)),
```
Comment (mkhokh-33) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java:32]: we could create separate method here
getS3CopyConfig(config)

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierFactory.java [hunk 2]
```diff
@@ -9,24 +9,23 @@
...
-                             final DestinationSyncMode syncMode,
-                             final String schema,
-                             final String streamName,
-                             final AmazonS3 s3Client,
-                             final JdbcDatabase db,
-                             final S3DestinationConfig s3Config,
-                             final ExtendedNameTransformer nameTransformer,
-                             final SqlOperations sqlOperations)
+  protected StreamCopier create(String stagingFolder,
+                                String schema,
```
Comment (mkhokh-33) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierFactory.java:20]: miss final before args

**Review by mkhokh-33 — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java (hunk 1 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java]: Updated, thanks

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierFactory.java (hunk 2 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierFactory.java]: Updated, thanks

**Review by mkhokh-33 — APPROVED** [review]

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java [hunk 3]
```diff
@@ -28,7 +29,7 @@ public AirbyteMessageConsumer getConsumer(final JsonNode config,
         getDatabase(config),
         getSqlOperations(),
         getNameTransformer(),
-        getS3DestinationConfig(config),
+        getS3CopyConfig(config),
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java:32]: I suspect you need to use `getS3CopyConfig(config.get("loading_method"))` here. But it's worth running acceptance tests to verify that.

File: airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  @Test
+  public void copiesCorrectFilesToTable() throws Exception {
+    // Generate two files
+    for (int i = 0; i < MAX_PARTS_PER_FILE + 1; i++) {
+      copier.prepareStagingFile();
+    }
+
+    copier.copyStagingFileToTemporaryTable();
+
+    verify(db).execute(String.format("COPY INTO fake-schema.%s FROM "
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierTest.java:80]: You should also verify that it copies the first file (I think this is only verifying that the second file is copied).

**Review by edgao — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/destination-snowflake/README.md [hunk 5]
```diff
@@ -22,7 +22,7 @@
 Put the contents of the `Snowflake Integration Test Config` secret on Rippling under the `Engineering` folder into `secrets/config.json` to be able to run integration tests locally.
 
 1. Put the contents of the `destination snowflake - insert test creds` LastPass secret into `secrets/insert_config.json`.
-1. Put the contents of the `destination snowflake - insert staging test creds` secret into `insert_staging_config.json`.
+1. Put the contents of the `destination snowflake - insert staging test creds` secret into `internal_staging_config.json`.
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-snowflake/README.md:25]: Should this be `secrets/internal_staging_config`?

**Review by tuliren — APPROVED** [review]

File: airbyte-integrations/connectors/destination-snowflake/README.md (hunk 5 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/README.md:25]: Have no idea :) I just aligned documentation to what we actually have a codebase and secret manager

File: airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierTest.java (hunk 4 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierTest.java:80]: Hi @edgao. Could you please clarify a little bit more about this point?

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopier.java [hunk 6]
```diff
@@ -47,4 +86,14 @@ public void copyS3CsvFileIntoTable(final JdbcDatabase database,
     database.execute(copyQuery);
   }
 
+  @VisibleForTesting
+  String getTmpTableName() {
```
[the remark this answers is not in the document]
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopier.java]: Updated, thanks

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java [hunk 7]
```diff
@@ -58,6 +59,10 @@ private String getConfiguredSchema(final JsonNode config) {
     return config.get("schema").asText();
   }
 
+  private S3CopyConfig getS3CopyConfig(final JsonNode config) {
```
[the remark this answers is not in the document]
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java]: Updated, thanks

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java (hunk 3 above)
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeCopyS3Destination.java]: Updated, thanks

File: airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierTest.java (hunk 4 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-snowflake/src/test/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StreamCopierTest.java:80]: The for-loop of `copier.prepareStagingFile()` is causing `copier` to create 2 files on S3. So we should verify that `db.execute("COPY INTO... FROM s3://.....")`  is being called for both of those files. Currently this test only checks that the second file is COPY-ed.

**Review by edgao — APPROVED** [review]
nice! :shipit:

**Pull request merged**